src/main/kotlin/org/jitsi/jicofo/jibri/JibriRecorder.kt

```diff
@@ -0,0 +1,198 @@
+ */
+package org.jitsi.jicofo.jibri
+
+import org.apache.commons.lang3.StringUtils
+import org.jitsi.impl.protocol.xmpp.XmppProvider
+import org.jitsi.jicofo.JitsiMeetConferenceImpl
+import org.jitsi.jicofo.jibri.JibriConfig.Companion.config
+import org.jitsi.jicofo.jibri.JibriSession.StartException
+import org.jitsi.jicofo.jibri.JibriSession.StartException.AllBusy
+import org.jitsi.jicofo.jibri.JibriSession.StartException.NotAvailable
+import org.jitsi.utils.logging2.Logger
+import org.jitsi.xmpp.extensions.jibri.JibriIq
+import org.jitsi.xmpp.extensions.jibri.JibriIq.RecordingMode
+import org.jitsi.xmpp.extensions.jibri.JibriIq.Status
+import org.jitsi.xmpp.extensions.jibri.RecordingStatus
+import org.jivesoftware.smack.packet.IQ
+import org.jivesoftware.smack.packet.XMPPError
+import org.jitsi.jicofo.util.ErrorResponse.create as error
+
+/**
+ * Handles conference recording through Jibri.
+ * Waits for updates from [JibriDetector] about recorder instance
+ * availability and publishes that information in Jicofo's MUC presence.
+ * Handles incoming Jibri IQs coming from conference moderator to
+ * start/stop the recording.
+ *
+ * @author Pawel Domas
+ * @author Sam Whited
+ */
+class JibriRecorder(
+    conference: JitsiMeetConferenceImpl,
+    xmppProvider: XmppProvider,
+    jibriDetector: JibriDetector,
+    parentLogger: Logger
+) : BaseJibri(
+    conference,
+    xmppProvider,
+    parentLogger,
+    jibriDetector
+) {
+    /**
+     * The current recording session or <tt>null</tt>.
+     */
+    private var jibriSession: JibriSession? = null
+
+    /**
+     * {@inheritDoc}
+     */
+    override fun dispose() {
+        jibriSession?.stop(null)
+        super.dispose()
+    }
+
+    /**
+     * Accepts only [JibriIq] without a SIP address.
+     */
+    override fun acceptType(packet: JibriIq): Boolean {
+        // the packet cannot contain a SIP address (must be handled
+        // by JibriSipGateway)
+        return StringUtils.isBlank(packet.sipAddress)
+    }
+
+    /**
+     * [JibriRecorder] has at most one jibri session.
+     */
+    override fun getJibriSessionForMeetIq(iq: JibriIq): JibriSession? = jibriSession
+
+    /**
+     * {@inheritDoc}
+     */
+    override fun getJibriSessions(): List<JibriSession> = jibriSession?.let { listOf(it) } ?: emptyList()
+
+    /**
+     * Handles a request to start a jibri session coming from a client in the conference and returns the response to
+     * be sent.
+     *
+     * [BaseJibri] has checked that there is no recording session currently active.
+     */
+    override fun handleStartRequest(iq: JibriIq): IQ {
+        val streamIdIsEmpty = StringUtils.isBlank(iq.streamId)
+        return if ((
+            !streamIdIsEmpty &&
+                iq.recordingMode == RecordingMode.STREAM ||
+                iq.recordingMode == RecordingMode.UNDEFINED
+            ) ||
+            streamIdIsEmpty && iq.recordingMode == RecordingMode.FILE
+        ) {
```

Comment:
Ugh, is this really the format that ktlint insists on? Any way it can be closed up at all?